File: reference-python/backend/planners/rrt_star.py
"""RRT* — RRT'nin yolu optimize eden hali."""

import time
from typing import Sequence

from backend.planners.base import Obstacle, PlanResult, Point
from backend.planners.rrt import RRTPlanner, _distance


class RRTStarPlanner(RRTPlanner):
    """Komsu yeniden baglama (rewire) ile yol maliyetini dusuren RRT varyanti."""

    name = "rrt_star"

    def __init__(self, collision_checker, rewire_radius: float = 0.15, **kwargs) -> None:
        super().__init__(collision_checker, **kwargs)
        self.rewire_radius = rewire_radius

    def plan(self, start: Point, goal: Point, obstacles: Sequence[Obstacle]) -> PlanResult:
        started_at = time.perf_counter()

        if not self.collision_checker(start) or not self.collision_checker(goal):
            return PlanResult(
                success=False, elapsed_seconds=time.perf_counter() - started_at, algorithm=self.name
            )

        bounds = self._workspace_bounds(start, goal)
        nodes = [start]
        parent = {start: None}
        cost = {start: 0.0}
        children = {start: set()}
        nodes_expanded = 0
        goal_candidates = []

        # Hedefe ulasan ilk dugumden sonra da aramaya devam ediyoruz; RRT'nin
        # aksine amac sadece bir yol degil, rewire ile giderek kisalan bir yol.
        for _ in range(self.max_iterations):
            nodes_expanded += 1
            sample = goal if self._rng.random() < self.goal_bias else self._sample(bounds)
            nearest = self._nearest(nodes, sample)
            new_point = self._steer(nearest, sample)

            if new_point in cost:
                continue
            if not self.collision_checker(new_point) or not self._segment_free(nearest, new_point):
                continue

            neighbors = [n for n in nodes if _distance(n, new_point) <= self.rewire_radius] or [nearest]

            best_parent = nearest
            best_cost = cost[nearest] + _distance(nearest, new_point)
            for neighbor in neighbors:
                candidate_cost = cost[neighbor] + _distance(neighbor, new_point)
                if candidate_cost < best_cost and self._segment_free(neighbor, new_point):
                    best_parent = neighbor
                    best_cost = candidate_cost

            nodes.append(new_point)
            parent[new_point] = best_parent
            cost[new_point] = best_cost
            children[new_point] = set()
            children[best_parent].add(new_point)

            for neighbor in neighbors:
                if neighbor == best_parent:
                    continue
                candidate_cost = best_cost + _distance(new_point, neighbor)
                if candidate_cost < cost[neighbor] and self._segment_free(new_point, neighbor):
                    previous_cost = cost[neighbor]
                    previous_parent = parent[neighbor]
                    if previous_parent is not None:
                        children[previous_parent].discard(neighbor)
                    parent[neighbor] = new_point
                    children[new_point].add(neighbor)
                    _propagate_descendant_costs(
                        neighbor, candidate_cost - previous_cost, children, cost
                    )

            if _distance(new_point, goal) <= self.goal_tolerance:
                if new_point == goal or self._segment_free(new_point, goal):
                    goal_candidates.append(new_point)

        best_goal_node = min(
            goal_candidates,
            key=lambda node: cost[node] + _distance(node, goal),
            default=None,
        )

        if best_goal_node is None:
            return PlanResult(
                success=False,
                path=[],
                elapsed_seconds=time.perf_counter() - started_at,
                nodes_expanded=nodes_expanded,
                algorithm=self.name,
            )

        path = self._build_path(parent, best_goal_node, start, goal)
        return PlanResult(
            success=True,
            path=path,
            elapsed_seconds=time.perf_counter() - started_at,
            nodes_expanded=nodes_expanded,
            algorithm=self.name,
        )


def _propagate_descendant_costs(root, delta, children, cost) -> None:
    """Rewire maliyet farkini kok ve tum torunlarina uygular."""
    pending = [root]
    while pending:
        node = pending.pop()
        cost[node] += delta
        pending.extend(children[node])

File: lib/robotics/planners/rrtStar.ts
import type { Vec3 } from "../transform";
import type { CollisionChecker, PlanResult } from "./base";
import { distance } from "./base";
import { RrtPlanner, type RrtOptions } from "./rrt";

export interface RrtStarOptions extends RrtOptions {
  /** Yeniden bağlama (rewire) için komşu arama yarıçapı. */
  rewireRadius?: number;
}

export function propagateDescendantCosts(
  root: Vec3,
  delta: number,
  children: Map<Vec3, Set<Vec3>>,
  cost: Map<Vec3, number>,
): void {
  const pending = [root];
  while (pending.length > 0) {
    const node = pending.pop()!;
    cost.set(node, (cost.get(node) ?? 0) + delta);
    for (const child of children.get(node) ?? []) pending.push(child);
  }
}

/**
 * RRT* — komşu yeniden bağlama (rewire) ile yol maliyetini düşüren RRT
 * varyantı. reference-python/backend/planners/rrt_star.py'nin TS portu.
 * Hedefe ulaşan ilk düğümden sonra da arama devam eder; RRT'nin aksine amaç
 * sadece bir yol değil, rewire ile giderek kısalan bir yoldur.
 */
export class RrtStarPlanner extends RrtPlanner {
  name = "rrt_star";
  private rewireRadius: number;

  constructor(options: RrtStarOptions = {}) {
    super(options);
    this.rewireRadius = options.rewireRadius ?? 0.15;
  }

  plan(start: Vec3, goal: Vec3, isFree: CollisionChecker): PlanResult {
    const startedAt = Date.now();
    if (!isFree(start) || !isFree(goal)) {
      return { success: false, path: [], elapsedMs: Date.now() - startedAt, nodesExpanded: 0, algorithm: this.name };
    }

    const bounds = this.workspaceBounds(start, goal);
    const nodes: Vec3[] = [start];
    const parent = new Map<Vec3, Vec3 | null>([[start, null]]);
    const children = new Map<Vec3, Set<Vec3>>([[start, new Set()]]);
    const cost = new Map<Vec3, number>([[start, 0]]);
    let nodesExpanded = 0;
    const goalCandidates: Vec3[] = [];

    for (let i = 0; i < this.maxIterations; i++) {
      nodesExpanded++;
      const sample = this.random() < this.goalBias ? goal : this.sample(bounds);
      const nearest = this.nearest(nodes, sample);
      const newPoint = this.steer(nearest, sample);

      if (cost.has(newPoint)) continue;
      if (!isFree(newPoint) || !this.segmentFree(nearest, newPoint, isFree)) continue;

      const nearby = nodes.filter((n) => distance(n, newPoint) <= this.rewireRadius);
      const neighbors = nearby.length > 0 ? nearby : [nearest];

      let bestParent = nearest;
      let bestCost = (cost.get(nearest) ?? 0) + distance(nearest, newPoint);
      for (const neighbor of neighbors) {
        const candidateCost = (cost.get(neighbor) ?? 0) + distance(neighbor, newPoint);
        if (candidateCost < bestCost && this.segmentFree(neighbor, newPoint, isFree)) {
          bestParent = neighbor;
          bestCost = candidateCost;
        }
      }

      nodes.push(newPoint);
      parent.set(newPoint, bestParent);
      cost.set(newPoint, bestCost);
      children.set(newPoint, new Set());
      children.get(bestParent)?.add(newPoint);

      for (const neighbor of neighbors) {
        if (neighbor === bestParent) continue;
        const candidateCost = bestCost + distance(newPoint, neighbor);
        if (candidateCost < (cost.get(neighbor) ?? Infinity) && this.segmentFree(newPoint, neighbor, isFree)) {
          const previousCost = cost.get(neighbor) ?? Infinity;
          const previousParent = parent.get(neighbor);
          if (previousParent) children.get(previousParent)?.delete(neighbor);
          parent.set(neighbor, newPoint);
          children.get(newPoint)?.add(neighbor);
          propagateDescendantCosts(neighbor, candidateCost - previousCost, children, cost);
        }
      }

      // RRT'deki ile aynı kural: hedefe yakın olmak yetmez, son sıçrama da
      // çarpışmasız olmalı. RRT*'de bu düğüm yalnızca aday olarak
      // saklandığı için kontrolü burada yapıyoruz — aksi hâlde kontrol
      // edilmemiş bir son segment en iyi yol olarak seçilebilirdi.
      if (distance(newPoint, goal) <= this.goalTolerance) {
        const sonSegmentGuvenli = newPoint === goal || this.segmentFree(newPoint, goal, isFree);
        if (sonSegmentGuvenli) goalCandidates.push(newPoint);
      }
    }

    const bestGoalNode = goalCandidates.reduce<Vec3 | null>((best, candidate) => {
      if (best === null) return candidate;
      const candidateCost = (cost.get(candidate) ?? Infinity) + distance(candidate, goal);
      const bestCost = (cost.get(best) ?? Infinity) + distance(best, goal);
      return candidateCost < bestCost ? candidate : best;
    }, null);

    if (bestGoalNode === null) {
      return { success: false, path: [], elapsedMs: Date.now() - startedAt, nodesExpanded, algorithm: this.name };
    }

    const path = this.buildPath(parent, bestGoalNode, goal);
    return { success: true, path, elapsedMs: Date.now() - startedAt, nodesExpanded, algorithm: this.name };
  }
}
